### PR TITLE
matrix ops tests & benchmarks

### PR DESCRIPTION
**Pull request by baykaner**
- BooleanMask: Fix, test & benchmark
- Scatter: Fix, test & benchmark
- Max: test & benchmark
- Tensor major order flip fix
- implemented math numeric limits calls
- Scatter1d, Scatter2d, Scatter3d: implemented, test & benchmark
- Min: benchmarking
- Maximum: fix, test & benchmarked

### CODE REVIEW

File: libs/math/include/math/matrix_operations.hpp [hunk 1]
```diff
@@ -115,69 +118,139 @@ meta::IfIsMathArray<ArrayType, void> BooleanMask(ArrayType const &input_array,
...
+ * @param input_array the input tensor to update
+ * @param updates the update values to apply
+ * @param indices vector of indices at which to apply the updates
  */
-namespace details {
-template <typename ArrayType>
-void ScatterImplementation(ArrayType &input_array, ArrayType &updates, ArrayType &indices)
+template <typename ArrayType, typename... Indices>
+void Scatter1D(ArrayType &input_array, ArrayType const &updates,
+               std::vector<Indices...> const &indices)
```
Comment (pierre-wilmot-fetch-ai) [inline libs/math/include/math/matrix_operations.hpp]: This looks very strange, either use a vector of uints or a parameter pack

File: libs/math/include/math/matrix_operations.hpp [hunk 2]
```diff
@@ -115,69 +118,139 @@ meta::IfIsMathArray<ArrayType, void> BooleanMask(ArrayType const &input_array,
...
+ * @tparam ArrayType Tensor of some data type
+ * @tparam Indices indices which are to be updated
+ * @param input_array the input tensor to update
+ * @param updates the update values to apply
+ * @param indices_0 the indices at which to apply the updates for the 0th dimension
+ * @param indices_1 the indices at which to apply the updates for the 1st dimension
+ */
+template <typename ArrayType, typename... Indices>
+void Scatter2D(ArrayType &input_array, ArrayType const &updates,
+               std::vector<Indices...> const &indices_0, std::vector<Indices...> const &indices_1)
```
Comment (pierre-wilmot-fetch-ai) [inline libs/math/include/math/matrix_operations.hpp]: Same here

File: libs/math/include/math/matrix_operations.hpp [hunk 3]
```diff
@@ -115,69 +118,139 @@ meta::IfIsMathArray<ArrayType, void> BooleanMask(ArrayType const &input_array,
...
+ * @tparam Indices indices which are to be updated
+ * @param input_array the input tensor to update
+ * @param updates the update values to apply
+ * @param indices_0 the indices at which to apply the updates for the 0th dimension
+ * @param indices_1 the indices at which to apply the updates for the 1st dimension
+ */
+template <typename ArrayType, typename... Indices>
+void Scatter3D(ArrayType &input_array, ArrayType const &updates,
+               std::vector<Indices...> const &indices_0, std::vector<Indices...> const &indices_1,
+               std::vector<Indices...> const &indices_2)
```
Comment (pierre-wilmot-fetch-ai) [inline libs/math/include/math/matrix_operations.hpp]: same here

File: libs/math/include/math/matrix_operations.hpp [hunk 4]
```diff
@@ -115,69 +118,139 @@ meta::IfIsMathArray<ArrayType, void> BooleanMask(ArrayType const &input_array,
...
+{
+  ASSERT(input_array.shape().size() == 3);
+  ASSERT(indices_0.size() == updates.size());
+  ASSERT(indices_1.size() == updates.size());
+  ASSERT(indices_2.size() == updates.size());
+
+  auto indices_it_0 = indices_0.begin();
+  auto indices_it_1 = indices_1.begin();
+  auto indices_it_2 = indices_2.begin();
+  for (auto &update_val : updates)
```
Comment (pierre-wilmot-fetch-ai) [inline libs/math/include/math/matrix_operations.hpp]: const &

File: libs/math/include/math/matrix_operations.hpp [hunk 5]
```diff
@@ -115,69 +118,139 @@ meta::IfIsMathArray<ArrayType, void> BooleanMask(ArrayType const &input_array,
...
+{
+  ASSERT(input_array.shape().size() == 2);
+  ASSERT(indices_0.size() == updates.size());
+  ASSERT(indices_1.size() == updates.size());
 
-  // Place back into arrays
-  for (size_t i = 0; i < updates.size(); ++i)
+  auto indices_it_0 = indices_0.begin();
+  auto indices_it_1 = indices_1.begin();
+  for (auto &update_val : updates)
```
Comment (pierre-wilmot-fetch-ai) [inline libs/math/include/math/matrix_operations.hpp]: const &

File: libs/math/include/math/matrix_operations.hpp [hunk 6]
```diff
@@ -115,69 +118,139 @@ meta::IfIsMathArray<ArrayType, void> BooleanMask(ArrayType const &input_array,
...
-  // sort indices and updates into ascending order
-  std::vector<std::pair<typename ArrayType::SizeType, typename ArrayType::Type>> AB;
+  ASSERT(input_array.shape().size() == 1);
+  ASSERT(indices.size() == updates.size());
 
-  // copy into pairs
-  // Note that A values are put in "first" this is very important
-  for (typename ArrayType::SizeType i = 0; i < updates.size(); ++i)
+  auto indices_it = indices.begin();
+  for (auto &update_val : updates)
```
Comment (pierre-wilmot-fetch-ai) [inline libs/math/include/math/matrix_operations.hpp]: const &

File: libs/math/include/math/matrix_operations.hpp [hunk 7]
```diff
@@ -115,69 +118,139 @@ meta::IfIsMathArray<ArrayType, void> BooleanMask(ArrayType const &input_array,
...
+template <typename ArrayType, typename T, typename = std::enable_if_t<meta::IsArithmetic<T>>>
+meta::IfIsMathArray<ArrayType, void> Product(ArrayType const &array1, T &ret)
 {
-  ASSERT(updates.size() == indices.size());
-  ASSERT(input_array.size() >= Max(indices));
-
-  typename ArrayType::SizeType idx{0};
-  for (auto &update_val : updates)
+  ret = typename ArrayType::Type(1);
+  for (auto &val : array1)
```
Comment (pierre-wilmot-fetch-ai) [inline libs/math/include/math/matrix_operations.hpp]: const &

File: libs/math/include/math/matrix_operations.hpp [hunk 8]
```diff
@@ -115,69 +118,139 @@ meta::IfIsMathArray<ArrayType, void> BooleanMask(ArrayType const &input_array,
...
+/**
+ * return the product of all elements in the vector
+ * @tparam T
+ * @param obj1
+ * @param ret
+ */
+template <typename T>
+void Product(std::vector<T> const &obj1, T &ret)
+{
+  ret = std::accumulate(std::begin(obj1), std::end(obj1), std::size_t(1), std::multiplies<>());
```
Comment (pierre-wilmot-fetch-ai) [inline libs/math/include/math/matrix_operations.hpp]: std::size_t(1)  should probably be T(1)

**Review by pierre-wilmot-fetch-ai — CHANGES_REQUESTED** [review]

File: libs/math/include/math/matrix_operations.hpp (hunk 1 above)
Comment (baykaner) [reply libs/math/include/math/matrix_operations.hpp]: good point, I'll have to figure out how to do the requisite nested parameter packing

**Review by MissingNO57 — DISMISSED** [review]
Code looks good to me.

File: libs/math/include/math/base_types.hpp [hunk 9]
```diff
@@ -21,6 +21,9 @@
 #include <unordered_set>
 #include <vector>
 
+#include "core/vector.hpp"
```
Comment (pierre-wilmot-fetch-ai) [inline libs/math/include/math/base_types.hpp]: Seems unused

File: libs/math/include/math/matrix_operations.hpp [hunk 10]
```diff
@@ -115,67 +118,78 @@ meta::IfIsMathArray<ArrayType, void> BooleanMask(ArrayType const &input_array,
...
 
-  // copy into pairs
-  // Note that A values are put in "first" this is very important
-  for (typename ArrayType::SizeType i = 0; i < updates.size(); ++i)
+  auto     indices_it = indices.begin();
+  SizeType update_idx{0};
+  while (indices_it != indices.end())
   {
-    AB.push_back(std::make_pair(indices[i], updates[i]));
+    input_array[input_array.ComputeIndex(*indices_it)] = updates[update_idx];
```
Comment (pierre-wilmot-fetch-ai) [inline libs/math/include/math/matrix_operations.hpp:145]: input_array.at(*indices_it) = updates[update_idx]; // Would look cleaner 
Also, do we offer external access to tensor memory through operator[] without any consideration for padding ?

File: libs/math/include/math/matrix_operations.hpp [hunk 11]
```diff
@@ -115,67 +118,78 @@ meta::IfIsMathArray<ArrayType, void> BooleanMask(ArrayType const &input_array,
...
+    ++update_idx;
   }
+}
 
-  std::sort(AB.begin(), AB.end());
-
-  // Place back into arrays
-  for (size_t i = 0; i < updates.size(); ++i)
+/**
+ * returns the product of all values in one array - calls SIMD implementation
```
Comment (pierre-wilmot-fetch-ai) [inline libs/math/include/math/matrix_operations.hpp]: calls SIMD implementation // That looks inacurate judging from the code just below

File: libs/math/include/math/matrix_operations.hpp [hunk 12]
```diff
@@ -115,67 +118,78 @@ meta::IfIsMathArray<ArrayType, void> BooleanMask(ArrayType const &input_array,
...
+ * @tparam ArrayType
+ * @tparam T
+ * @param array1
+ * @param ret
+ */
+template <typename ArrayType, typename T, typename = std::enable_if_t<meta::IsArithmetic<T>>>
+meta::IfIsMathArray<ArrayType, void> Product(ArrayType const &array1, T &ret)
+{
+  ret = typename ArrayType::Type(1);
```
Comment (pierre-wilmot-fetch-ai) [inline libs/math/include/math/matrix_operations.hpp]: Also should have a check for the edge case when array is empty

File: libs/math/include/math/matrix_operations.hpp [hunk 13]
```diff
@@ -115,67 +118,78 @@ meta::IfIsMathArray<ArrayType, void> BooleanMask(ArrayType const &input_array,
...
 {
-  ASSERT(updates.size() == indices.size());
-  ASSERT(input_array.size() >= Max(indices));
-
-  typename ArrayType::SizeType idx{0};
-  for (auto &update_val : updates)
-  {
-    input_array.At(indices.At(idx)) = update_val;
-  }
+  ret = std::accumulate(std::begin(obj1), std::end(obj1), T(1), std::multiplies<>());
```
Comment (pierre-wilmot-fetch-ai) [inline libs/math/include/math/matrix_operations.hpp]: Handle empty vector case (return 0)

File: libs/math/include/math/tensor.hpp [hunk 14]
```diff
@@ -173,6 +173,7 @@ class Tensor
   SelfType &      FillUniformRandomIntegers(int64_t const &min, int64_t const &max);
   static SelfType Zeroes(SizeVector const &shape);
   static SelfType Ones(SizeVector const &shape);
+  SizeType        ComputeIndex(SizeVector const &indices);
```
Comment (pierre-wilmot-fetch-ai) [inline libs/math/include/math/tensor.hpp]: Method should be const

File: libs/math/tests/math/matrix_operations/matrix_operations.cpp [hunk 15]
```diff
@@ -36,6 +38,318 @@ using MyTypes = ::testing::Types<fetch::math::Tensor<float>, fetch::math::Tensor
...
+}
+
+TYPED_TEST(FreeFunctionsTest, Scatter1D_SetAll)
+{
+  using Type = typename TypeParam::Type;
+
+  TypeParam               array1{4};
+  TypeParam               updates{4};
+  std::vector<SizeVector> indices{};
+  updates.SetAllOne();
```
Comment (pierre-wilmot-fetch-ai) [inline libs/math/tests/math/matrix_operations/matrix_operations.cpp:64]: Not that important, but would be nice testing with different values.

**Review by pierre-wilmot-fetch-ai — DISMISSED** [review]

File: libs/math/include/math/matrix_operations.hpp (hunk 10 above)
Comment (baykaner) [reply libs/math/include/math/matrix_operations.hpp:145]: input_array.at(*indices_it) = updates[update_idx]
would require an at(vector) method and I'm reluctant to to implement an inefficient at(vector) or set(vector) method in tensor.
We do expose access to the underlying data this way, and when we handle padding in a better way the ComputeIndex method will likely need an update to consider this.

**Review by pierre-wilmot-fetch-ai — APPROVED** [review]

**Pull request merged**